Adding intersectional bias mitigation to AIF360

ckalousi opened the pull request:
We have implemented an intersectional bias mitigation algorithm based on https://doi.org/10.1007/978-3-030-87687-6_5 (see also https://doi.org/10.48550/arXiv.2010.13494 for the arxiv version) as discussed further on issue https://github.com/Trusted-AI/AIF360/issues/537.  Additional details are available in the demo notebook.

ckalousi commented:
According to pytest, our code does not reach the desired coverage of 80%.  This happens because our code is multi threaded and it was not obvious to us how pytest can support such kind of code.  Nonetheless, we have checked that all functions of the main algorithm file are called during our tests.

ckalousi commented:
@RahulVadisetty91 

Dear Rahul, 

we are very grateful for taking the time to review our code and offer some very valuable comments.

We are also extremely sorry we couldn't address your comments earlier as we had to observe some critical deadlines in our work and also coordinate our actions regarding our pull request.

About your comment on the init method.  This is a great catch.  Indeed self.model is not needed in the current setting.  We had only put it there in case we wanted in the future to expand support of our Intersectional Fairness to more algorithms.  Under current circumstances it makes sense to comment out this line (line 27 of your screenshot).

You have also made some very valuable comments in your first version of your comment (before the edit).  We considered all of them, and although they are all very important we could only use current resources to address some of them.

More specifically, although it would be nice to switch to TensorFlow 2 for future compatibility, our code is based on the code of the Adversarial Debiasing algorithm found in AIF360 which in turn is based on TensorFlow 1.  It is very difficult to modify our code to support TensorFlow 2 the time Adversarial Debiasing uses TensorFlow 1.  If in the future the original algorithm is updated we would be happy to also update our code.

Following one of your comments we have now implemented evaluation progress bars in our algorithm.

Once again thank you for your time and we would be happy to further discuss any of your suggestions or concerns.

Best regards,
Chrysostomos

hoffmansc commented on aif360/algorithms/isf_helpers/inprocessing/adversarial_debiasing.py:
```diff
@@ -0,0 +1,121 @@
+        tf.compat.v1.reset_default_graph()
+        return ds_predict
+
+    def bias_predict(self, ds_train):
+        """
+        Model learning and prediction using AdversarialDebiasing of AIF360 without debias.
+
+        Parameters
+        ----------
+        ds_train : Dataset
+            Dataset for training and prediction
+
+        Returns
+        -------
+        ds_predict : numpy.ndarray
+            Predicted label
+        """
+        ikey = ds_train.protected_attribute_names[0]
+        priv_g = [{ikey: ds_train.privileged_protected_attributes[0]}]
+        upriv_g = [{ikey: ds_train.unprivileged_protected_attributes[0]}]
+        sess = tf.compat.v1.Session()
+        model = AD(
+            privileged_groups=priv_g,
+            unprivileged_groups=upriv_g,
+            scope_name='plain_classifier',
+            debias=False,
+            sess=sess,
+            num_epochs=self.options['num_epochs'],
+            batch_size=self.options['batch_size'])
+        model.fit(ds_train)
+        ds_predict = model.predict(ds_train)
+        sess.close()
+        tf.compat.v1.reset_default_graph()
+        return ds_predict
```
unused

ckalousi replied:
removed

hoffmansc commented on aif360/algorithms/isf_helpers/postprocessing/eq_odds.py:
why can't we use the built-in [`EqOddsPostprocessing`](https://aif360.readthedocs.io/en/latest/modules/generated/aif360.algorithms.postprocessing.EqOddsPostprocessing.html#aif360.algorithms.postprocessing.EqOddsPostprocessing) here?

ckalousi replied:
due to some compatibility issues we had with the built-in function we decided to remove this functionality from our work

hoffmansc commented on aif360/algorithms/isf_helpers/preprocessing/massaging.py:
it would be nice to have this as a full-fledged algorithm like `Reweighing`

ckalousi replied:
Indeed it would be nice.  Currently aif360 does not support the `Massaging `algorithm.  Although it would be nice to have it as an independent algorithm, we have not taken any action here.  Maybe this can be addressed in the future?

hoffmansc commented on aif360/algorithms/isf_helpers/preprocessing/massaging.py:
```diff
@@ -0,0 +1,80 @@
+        ds_train : Dataset
+            Training dataset
+        """
+        scale_orig = StandardScaler()
+        X = scale_orig.fit_transform(ds_train.features)
```
is this necessary inside `fit()`? can't the user just apply scaling before passing the dataset?

ckalousi replied:
We have removed the scaling method.  Now the user can apply scaling before passing the dataset.

hoffmansc commented on aif360/algorithms/isf_helpers/inprocessing/adversarial_debiasing.py:
is there any way we can just use the `AdversarialDebiasing` class directly instead of this wrapper? this doesn't seem to be doing much.

ckalousi replied:
Since isf needs to distinguish between pre/in/post processing, it needs a wrapper to abstract it.  It would be hard to remove the wrapper classes.

hoffmansc replied:
Can you explain more? Why can't we just compare the algorithm name? i.e., `if self.algorithm == "AdversarialDebiasing": self.approach_type == "InProcessing"`

ckalousi replied:
thank you very much for your comment, we have now deleted the wrapper class and use the AIF360 classes directly

hoffmansc commented on aif360/algorithms/isf_helpers/postprocessing/reject_option_based_classification.py:
same as before -- can we get rid of these wrapper classes?

ckalousi replied:
same comment as the previous comment

hoffmansc commented on aif360/algorithms/intersectional_fairness.py:
```diff
@@ -0,0 +1,1003 @@
+from aif360.algorithms.isf_helpers.postprocessing.reject_option_based_classification import RejectOptionClassification
+from aif360.algorithms.isf_helpers.postprocessing.equalized_odds_postprocessing import EqualizedOddsPostProcessing
+
+from logging import getLogger, StreamHandler, ERROR, Formatter
```
can we get rid of the debugging lines or if they're still useful just add a verbose flag?

ckalousi replied:
yes, we have now removed it

hoffmansc commented on tests/test_isf.py:
```diff
@@ -0,0 +1,301 @@
+    def __init__(self, methodName='runTest'):
+        super().__init__(methodName=methodName)
+
+    def _read_modelanswer(self, s_result_singleattr, s_result_combattr):
+        # load of model answer
+        ma_singleattr_bias = pd.read_csv(MODEL_ANSWER_PATH + s_result_singleattr, index_col=0)
+        ma_combattr_bias = pd.read_csv(MODEL_ANSWER_PATH + s_result_combattr, index_col=0)
+        return ma_singleattr_bias, ma_combattr_bias
+
+    def _comp_dataframe(self, df1, df2):
+        try:
+            assert_frame_equal(df1, df2)
+        except AssertionError:
+            return False
+        return True
```
unused

ckalousi replied:
removed

hoffmansc reviewed (changes requested):
Thanks for your patience here! My comments are mostly concerned with cleaning up the code and reducing redundancies but overall it's in pretty good shape. Good work!

ckalousi commented on aif360/algorithms/isf_helpers/isf_metrics/disparate_impact.py:
```diff
@@ -0,0 +1,165 @@
+        if bunbo_rate == 0:
+            return np.nan
+
+        return (bunshi_rate/bunbo_rate)
```
Yes, it is.  We have replaced our functions withy built-in ones.  As a result we have deleted our file disparate_impact.py.

ckalousi commented on aif360/algorithms/isf_helpers/isf_analysis/intersectional_bias.py:
```diff
@@ -0,0 +1,161 @@
+            name += k + " = " + str(v) + ","
+        df_result.loc[name[:-1]] = di
+
+    return df_result
```
we have replaced with built-in functions

ckalousi commented on aif360/algorithms/isf_helpers/isf_metrics/disparate_impact.py:
```diff
@@ -0,0 +1,165 @@
+import matplotlib.cm as cm
+
+
+class DisparateImpact():
```
we have deleted the file and replaced with built-in functions

ckalousi commented on aif360/algorithms/isf_helpers/isf_analysis/intersectional_bias.py:
```diff
@@ -0,0 +1,161 @@
+        df_result.loc[protect_attr[0]+"="+str(val0)] = tmp_li
+    df_result = df_result.set_axis(col_list, axis=1)
+
+    return df_result
```
Yes, we have removed the method `calc_intersectionalbias_matrix()` and we now use the built-in functions of aif360 (the only place that was used was in the tutorial, but now we use aif360 in the tutorial)

ckalousi commented on aif360/algorithms/isf_helpers/isf_analysis/metrics.py:
We have replaced with built-in functions.  The file is no longer needed and we have deleted it.

ckalousi commented on aif360/algorithms/isf_helpers/isf_utils/common.py:
```diff
@@ -0,0 +1,442 @@
+        label_list = list(itertools.product(labelss[0], labelss[1], labelss[2], labelss[3], labelss[4]))
+    else:
+        raise ValueError(
+            "Up to 5 protected_attribute_names can be set.")
```
we agree and we have followed your suggestion

ckalousi commented on aif360/algorithms/intersectional_fairness.py:
```diff
@@ -0,0 +1,1003 @@
+        self.options = options
+        self.options['metric'] = metric
+        self.options['threshold'] = 0
+        algo = globals()[self.algorithm](options)
```
we agree and we have aligned with your comment

ckalousi commented on aif360/algorithms/intersectional_fairness.py:
```diff
@@ -0,0 +1,1003 @@
+
+        # df_result.to_csv(self.TO.out_dir + '/valid_result_stat.csv')
+        self.logger.debug('done.')
+        self.df_result = df_result
```
indeed, it now returns `self`

ckalousi commented on aif360/algorithms/intersectional_fairness.py:
```diff
@@ -0,0 +1,1003 @@
+                       TPR,
+                       TNR,
+                       bal_acc,
+                       precision,
+                       f1,
+                       m_sg_mitig.selection_rate(privileged=True),
+                       difference,
+                       ratio]
```
we have removed the unnecessary metrics

ckalousi commented on aif360/algorithms/intersectional_fairness.py:
```diff
@@ -0,0 +1,1003 @@
+            ds_target.labels[fav_instances_idx] = ds_target.favorable_label
+            ds_target.labels[ufav_instances_idx] = ds_target.unfavorable_label
+        except ValueError:
+            exit()
```
yes, we have now removed it

ckalousi commented on aif360/algorithms/intersectional_fairness.py:
```diff
@@ -0,0 +1,1003 @@
+                    disable_df = pd.concat([disable_df, sdf])
+
+        unprivileged_protected_attributes_keys = self._get_attribute_keys(dataset, unprivileged_protected_attributes)
+        privileged_protected_attributes_keys = self._get_attribute_keys(dataset, privileged_protected_attributes)
```
true, we have now removed them

ckalousi commented on aif360/algorithms/intersectional_fairness.py:
```diff
@@ -0,0 +1,1003 @@
+                    if key1 == key2:
+                        attributes_keys.append(np.array([float(item[key2])]))
+                        break
+        return attributes_keys
```
unused, we have removed

ckalousi commented on aif360/algorithms/intersectional_fairness.py:
```diff
@@ -0,0 +1,1003 @@
+            v_mitig_target = vars(ds_mitig_target_pair)
+            with open(self.TO.out_dir + '/ds_mitig_target_pair.txt', 'w') as f:
+                self.logger.error(v_mitig_target, file=f)
+            self.logger.error(traceback.format_exc())
```
this part of the code was not necessary and we have now removed it

ckalousi commented on aif360/algorithms/intersectional_fairness.py:
```diff
@@ -0,0 +1,1003 @@
+
+        # Stores fav confidence (voting rate) for each instance
+        fav_voting_rate_dict = {}  # key:instance_name, value:fav_confidence
+        score_type = 3  # 1:only score, 2:avg(vote*score), 3:avg(vote)*w + avg(score)*(1-w)
```
we have removed `score_type`

ckalousi commented:
Dear @hoffmansc thank you very much for your patience while reviewing our code.  We agree to all of your comments and we have tried to address them to the best of our understanding.  We are committed to continue working with you, until we reach the required standards for publication.

ckalousi commented:
Dear maintainers, we have now addressed the requested changes to the best of our understanding and we are looking forward to your further comments at your convenience.  Thank you in advance.

hoffmansc commented on aif360/algorithms/intersectional_fairness.py:
```diff
@@ -0,0 +1,881 @@
+
+from tqdm import tqdm
+
+class IntersectionalFairness():
```
can we have this extend `aif360.algorithms.Transformer`?
```suggestion
class IntersectionalFairness(Transformer):
```

ckalousi replied:
indeed, thank you, we have now extended as suggested

ckalousi commented:
Dear @hoffmansc, thank you very much for your latest review.  We have now addressed all of your comments to the best of our understanding and we are waiting for further communication.  We hope we are reaching closer to publication.

ckalousi commented:
@hoffmansc Please let us know if there is anything we can amend on our end?

hoffmansc commented:
> @hoffmansc Please let us know if there is anything we can amend on our end?

I see one test related to this code:
FAILED tests/test_isf.py::TestStringMethods::test01_AdversarialDebiasing - AssertionError: DataFrame.iloc[:, 2] (column name="selection_rate") are different

The rest you can ignore.

ckalousi commented:
> > @hoffmansc Please let us know if there is anything we can amend on our end?
> 
> I see one test related to this code: FAILED tests/test_isf.py::TestStringMethods::test01_AdversarialDebiasing - AssertionError: DataFrame.iloc[:, 2] (column name="selection_rate") are different
> 
> The rest you can ignore.

@hoffmansc Thank you for your prompt and kind response.  We have made appropriate changes to the test file and successfully tested the code in our environment.  Please, let us know if the problem persists.

ckalousi commented:
@hoffmansc Thank you very much for your hard work and for assisting us in every step.

hoffmansc commented:
Of course, congrats for seeing it through. Sorry it took so long!
